### DISCUSSION TITLE
[SUGGESTION] Chronometer and timer logic blocks

### DISCUSSION THREAD

**Discussion by Uncle-Ulty** [tsunamayo/Starship-EVO#4034] [Suggestions]
Hey!

I'm building racing tracks for hovercraft race! And It would be awesome if we had a sort of chronometer or timer to register the lap time score! You know, to see who is the fastest pilot in the galaxy!

The first racing track is a MUTE CITY replica, from F-ZERO snes (I'm calling it "F-ZEVO :p)

I hope you enjoy it!, It's on workshop already. And I built the 3 original hovercrafts: Blue Falcon, Fire Stingray and Golden Fox!

youtube link:
https://www.youtube.com/watch?v=-0frSqlY0Hc&feature=youtu.be

workshop link
https://steamcommunity.com/sharedfiles/filedetails/?id=2368822900

![image](https://user-images.githubusercontent.com/73253472/105409181-d6ebb080-5c0e-11eb-927d-3973b7767bdc.png)

![image](https://user-images.githubusercontent.com/73253472/105409609-71e48a80-5c0f-11eb-840a-3c33a6585e8e.png)

**Comment by cloudrambler** [comment]
@Uncle-Ulty A chronometer could be built from existing game elements. I think that a timer gate driven latching register would work to about 0.2s consistency (I had a quick play with this). If you build one register to count in 0.2s steps up to 1s, a second register to count in 1s steps up to 10s, and a third register to count in 10s steps up to 1m. You could count minutes as well if you need to. Then each register would give you one value to convert into your display format. It would be possible to encode characters into a digital sign. If you want I can build an example.

**Comment by Uncle-Ulty** [comment]
I got the idea, but I'd require a 0,01s as minimum resolution, 0,2s is too much for racing

**Comment by Uncle-Ulty** [comment]
Also, it would be nice to connect a message log with the time, and somehow register some times, and display it. We already have some monitors for this.

**Comment by cloudrambler** [comment]
I think its possible to build a counter with a 0.01s resolution and I have made a prototype seven-segment display that is pretty good. But it will require a LOT of active parts and likely slow down the FPS, so there is no assurance that the time would be accurate.

Also there is no automatic trigger unless the sensor is able to detect different players; I've not tested that. So you would have to start/stop manually.

**Comment by tsunamayo** [comment]
Aha amazing! Yes, there is quite a few things needed first to make this works. Entity detection, custom text screen ect ect.

**Comment by Uncle-Ulty** [comment]
the entity detection we already have, I mean, player detection. I'm using a volumetric sensor to trigger a repeater. it activates due to the player inside the hovercraft, so it works. I needed this for, behold, the Rainbow Road (snes version). I need some trigger to initiate twomp movement. It's working pretty well. Already available in the workshop too, with a youtube video.

For de counter, I'm planning to make it with wait blocks. Ten 0,01 blocks, ten 0,1 blocks, ten 1s blocks and so on...

And I'd like to say this: with the already available tools in-game, there are tons of things we can build! I made the two first racing tracks and a loop for hovercraft. I also built a tube, where the hovercraft can hover inside in all directions, including upside down! this is amazing! This opens new possibilities for Starship EVO!

![image](https://user-images.githubusercontent.com/73253472/105571048-9f265b00-5d2b-11eb-9476-7e0d1c5c13e4.png)




![26](https://user-images.githubusercontent.com/73253472/105571105-e9a7d780-5d2b-11eb-9449-e9f4dfa142b2.png)

**Comment by cloudrambler** [comment]
@Uncle-Ulty I finished the display. It's on Steam if you want to use it. https://steamcommunity.com/sharedfiles/filedetails/?id=2372463143

I did not get very far with the timer and ran into some failed logic conditions (logic states that appear to be incorrect). However the method I was attempting is similar to what you have discussed.

**Comment by Uncle-Ulty** [comment]
thanks! gonna check it!

**Comment by Malus-Draco** [comment]
Timers are easy to make with SR Latches, i do agree though a clock block would make things easier for my project (im building a binary based ship computer, probably gonna be 8 or 16Bit). 

In the current experiment branch you can configure volumetric sensors to trigger on player, entity or item so obviously that would be your trigger to set your latch and you'd have a 0.5 second delay to reset the latch then another 0.5 delay set to the NOR output of the reset signal to set the latch again and start the clock cycle, I think what you need then is a counter to count the number of ticks, i would likely either use more latches or switches as well as AND gates (each one trigger when the last and the tick signal is high) for that to give 1's and 10's of seconds and finally 6 of 10's for minutes for as many minutes as you need maximum (im assuming you'll not need hours).
